library/core/src/array/mod.rs

```diff
@@ -463,6 +463,34 @@ impl<T, const N: usize> [T; N] {
         unsafe { crate::mem::transmute_copy::<_, [U; N]>(&dst) }
     }
 
+    /// 'Zips up' two arrays into a single array of pairs.
+    /// `zip()` returns a new array where every element is a tuple where the first element comes from the first array, and the second element comes from the second array.
+    /// In other words, it zips two arrays together, into a single one.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(array_zip)]
+    /// let x = [1, 2, 3];
+    /// let y = [4, 5, 6];
+    /// let z = x.zip(y);
+    /// assert_eq!(z, [(1, 4), (2, 5), (3, 6)]);
+    /// ```
+    #[unstable(feature = "array_zip", issue = "none")]
+    pub fn zip<U>(self, rhs: [U; N]) -> [(T, U); N] {
```

Comment:
One use case I had in mind when writing this PR was for something like a `struct Vector<const N: usize, T>([T; N])` to help with binary operations like
```rust
impl Add for Vector<const N: usize, T> where T: Add<T> {
    type Output = Vector<N, T>;
    fn add(self, rhs) -> ... {
        Vector(self.0.zip(rhs.0).map(|(lhs, rhs)| lhs + rhs)
    }
}
```

however would it make more sense to have a function which performed both the `map` and the `zip` as the same step without one extra intermediate array? Is this what @jplatte had in mind with `zip_with`?

Reply:
I didn't have a specific use case in mind when suggesting `zip_with`, but sure, that would be one.

Reply:
Interestingly enough, there is no `zip_with` for `Iterator` right now, only for `Option` ([link to docs](https://doc.rust-lang.org/nightly/std/?search=zip_with)). `zip_with` should maybe be done with a separate tracking issue and PR to add it in all the relevant places at once and see if there is even interest with a normal issue first.

Reply:
There seems to be an open issue for `zip_with` for `Iterator` #46654, however there does not seem to be that much interest. I guess I'll wait with array `zip_with` for now